src/Exception/HttpException.php

```diff
@@ -0,0 +1,212 @@
+<?php
+/**
+ * HTTP exception.
+ *
+ * HTTP exceptions are public-facing, so the messages are translated and should
+ * not reveal any sensitive information. Additional information can be attached
+ * for debugging.
+ *
+ * @package SatisPress
+ * @license GPL-2.0-or-later
+ * @since 0.3.0
+ */
+
+declare ( strict_types = 1 );
+
+namespace SatisPress\Exception;
+
+use SatisPress\Package;
+use SatisPress\Release;
+use Throwable;
+use WP_Http as HTTP;
+
+/**
+ * HTTP exception class.
+ *
+ * @since 0.3.0
+ */
+class HttpException extends \Exception implements ExceptionInterface {
+	/**
+	 * Exception data.
+	 *
+	 * @var array
+	 */
+	protected $data = [];
+
+	/**
+	 * HTTP status code.
+	 *
+	 * @var integer
+	 */
+	protected $status_code;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @param string    $message     Message.
+	 * @param integer   $status_code Optional. HTTP status code. Defaults to 500.
+	 * @param array     $data        Optional. Additional data.
+	 * @param integer   $code        Exception code.
+	 * @param Throwable $previous    Previous exception.
+	 */
+	public function __construct(
+		string $message,
+		int $status_code = HTTP::INTERNAL_SERVER_ERROR,
+		array $data = null,
+		int $code = 0,
+		Throwable $previous = null
+	) {
+		$this->data        = $data;
+		$this->status_code = $status_code;
+		$message           = $message ?: esc_html( 'Internal Server Error', 'satispress' );
+
+		parent::__construct( $message, $code, $previous );
+	}
+
+	/**
+	 * Create an exception for a forbidden resource request.
+	 *
+	 * @since 0.3.0.
+	 *
+	 * @param array     $data     Optional. Extra data for debugging.
+	 * @param int       $code     Optional. The Exception code.
+	 * @param Throwable $previous Optional. The previous throwable used for the exception chaining.
+	 * @return HTTPException
+	 */
+	public static function forForbiddenResource(
+		array $data = null,
+		int $code = 0,
+		Throwable $previous = null
+	): HTTPException {
+		$message = esc_html__( 'Sorry, you are not allowed to view this resource.', 'satispress' );
```

Comment:
As per https://www.alainschlesser.com/structuring-php-exceptions/, exception messages should only be seen...exceptionally. And even then, they are more likely to appear in log messages (which need consistency), than _intentionally_ displayed to Users.

Reply:
I've worked on a few projects in the past that use HTTP exceptions this way, but I'm thinking we might be able to do something a little different. The `getMessage()` method is final, so I put the display logic in a `displayMessage()` method for now, but that can go away at some point.

What I'm thinking is that when the response handling is fully abstracted, we can add a `toResponse()` method to the exception to convert it to an HTTP response. When debug mode is enabled, we can display the exception details, otherwise we'll show a public-facing message.